src/main/java/org/logstash/beats/BeatsParser.java

```diff
@@ -0,0 +1,230 @@
+import java.util.zip.InflaterInputStream;
+
+
+public class BeatsParser extends ByteToMessageDecoder {
+    private static final int CHUNK_SIZE = 1024;
+    private final static Logger logger = LogManager.getLogger(Server.class.getName());
+
+    private Batch batch = new Batch();
+
+    private enum States {
+        READ_HEADER,
+        READ_FRAME_TYPE,
+        READ_WINDOW_SIZE,
+        READ_JSON_HEADER,
+        READ_COMPRESSED_FRAME_HEADER,
+        READ_COMPRESSED_FRAME,
+        READ_JSON,
+        READ_DATA_FIELDS,
+    }
+
+    private States currentState = States.READ_HEADER;
+    private long requiredBytes = 0;
+    private int sequence = 0;
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        if(!hasEnoughBytes(in)) {
+            return;
+        }
+
+        switch (currentState) {
+            case READ_HEADER: {
+                logger.debug("Running: READ_HEADER");
+
+                byte currentVersion = in.readByte();
+
+                if(Protocol.isVersion2(currentVersion)) {
+                    logger.debug("Frame version 2 detected");
+                    batch.setProtocol(Protocol.VERSION_2);
+                } else {
+                    logger.debug("Frame version 1 detected");
+                    batch.setProtocol(Protocol.VERSION_1);
+                }
+
+                transition(States.READ_FRAME_TYPE, 1);
+                break;
+            }
+            case READ_FRAME_TYPE: {
+                byte frameType = in.readByte();
+
+                switch(frameType) {
+                    case Protocol.CODE_WINDOW_SIZE: {
+                        transition(States.READ_WINDOW_SIZE, 4);
+                        break;
+                    }
+                    case Protocol.CODE_JSON_FRAME: {
+                        // Reading Sequence + size of the payload
+                        transition(States.READ_JSON_HEADER, 8);
+                        break;
+                    }
+                    case Protocol.CODE_COMPRESSED_FRAME: {
+                        transition(States.READ_COMPRESSED_FRAME_HEADER, 4);
+                        break;
+                    }
+                    case Protocol.CODE_FRAME: {
+                        transition(States.READ_DATA_FIELDS, 8);
+                        break;
+                    }
+                }
+                break;
+            }
+            case READ_WINDOW_SIZE: {
+                logger.debug("Running: READ_WINDOW_SIZE");
+
+                batch.setBatchSize((int) in.readUnsignedInt());
+
+                // This is unlikely to happen but I have no way to known when a frame is
+                // actually completely done other than checking the windows and the sequence number,
+                // If the FSM read a new window and I have still
+                // events buffered I should send the current batch down to the next handler.
+                if(!batch.isEmpty()) {
```

Comment:
implementation detail question here: Should this actual return an error to the client as he potentially has done sth wrong? Or are we good to go and live our lifes?